runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -11360,6 +11360,364 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+   TR::Register *outputAddressReg = cg->gprClobberEvaluate(node->getChild(2));
+   TR::Register *outputOffsetReg = cg->gprClobberEvaluate(node->getChild(3));
+   TR::Register *remainingReg = cg->gprClobberEvaluate(node->getChild(4));
+
+   TR::Register *tempReg = inputOffsetReg;
+   TR::Register *maskReg = outputOffsetReg;
+
+   TR::Register *vec1Reg = cg->allocateRegister(TR_VRF);
+   TR::Register *vec2Reg = cg->allocateRegister(TR_VRF);
+   TR::Register *constZeroVecReg = cg->allocateRegister(TR_VRF);
+   TR::Register *constNeg16VecReg = (isLE && !isAtLeastP9) ? cg->allocateRegister(TR_VRF) : nullptr;
+
+   TR::Register *condReg = cg->allocateRegister(TR_CCR);
+
+   TR::LabelSymbol *startLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *load16LoopLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *load8Label = generateLabelSymbol(cg);
+   TR::LabelSymbol *load4Label = generateLabelSymbol(cg);
+   TR::LabelSymbol *load1Label = generateLabelSymbol(cg);
+   TR::LabelSymbol *doneLabel = generateLabelSymbol(cg);
+
+   int32_t numRegs = constNeg16VecReg ? 10 : 9;
+   TR::RegisterDependencyConditions *dependencies = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, numRegs, cg->trMemory());
+   dependencies->addPostCondition(inputAddressReg, TR::RealRegister::NoReg);
+   dependencies->getPostConditions()->getRegisterDependency(0)->setExcludeGPR0();
+   dependencies->addPostCondition(inputOffsetReg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(outputAddressReg, TR::RealRegister::NoReg);
+   dependencies->getPostConditions()->getRegisterDependency(2)->setExcludeGPR0();
+   dependencies->addPostCondition(outputOffsetReg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(remainingReg, TR::RealRegister::NoReg);
+   dependencies->getPostConditions()->getRegisterDependency(4)->setExcludeGPR0();
+   dependencies->addPostCondition(vec1Reg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(vec2Reg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(constZeroVecReg, TR::RealRegister::NoReg);
+   if (constNeg16VecReg)
+      {
+      dependencies->addPostCondition(constNeg16VecReg, TR::RealRegister::NoReg);
+      }
+   dependencies->addPostCondition(condReg, TR::RealRegister::cr0);
+
+   cg->generateDebugCounter(TR::DebugCounter::debugCounterName(comp, "inlineIntrinsicInflate/(%s)", comp->signature()));
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, startLabel, NULL);
+   startLabel->setStartInternalControlFlow();
+
+   /* If remainingReg equals 0, there is no work to be done. Immediate jump to the end. */
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpi4, node, condReg, remainingReg, 0);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::beq, node, doneLabel, condReg);
+
+   /* Add array header size and offset to the start of the array to determine the address of the first byte to read. */
```

Comment:
adjust these addresses differently for off-heap and i would expect the rest just works for it naturally. you currently do:  `addrReg = addrReg + arrayHeaderSize; `  for off-heap, you would do:
`           addrReg = loadDataAddrField(addrReg);`
especially, your enabling flag is checking canGenerateArraylets() which doesn't include off-heap anyway.

Reply:
Is `loadDataAddrField` the correct name? I'm not seeing anything with that name?

Reply:
the fieldname is dataAddr. there are quite some existing code doing its load. newArray needs to set it up as well.

Reply:
I changed this so that if `isIndexableDataAddrPresent` is true it will find dataAddr via `offsetOfContiguousDataAddrField` and load from it to find where the array data is instead of adding the header size. This seems to work when I tested it but please confirm I didn't overlook any details.

Reply:
@rmnattas Could you take a look at the updated version of this section and confirm that the correct checks are being done and that `TR_TARGET_64BIT` is the correct macro to use?

Reply:
You would want to use the `J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION` for builds that support off-heap. Also check using `TR::Compiler->om.isOffHeapAllocationEnabled()` instead of `TR::Compiler->om.isOffHeapAllocationEnabled()` ([example](https://github.com/eclipse-openj9/openj9/blob/74878931493e81abd35da8b3240c30ecf5a6b3ed/runtime/compiler/optimizer/J9TransformUtil.cpp#L94-L108)).

This is because you can have the dataAddr field existing in a non-off-heap run, by enabling dual header shape, but not used. Apparently it's possible because of prior work before deciding that the dataAddr field should only exist in off-heap.

Reply:
I have now changed `TR_TARGET_64BIT` to `J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION` and also `TR::Compiler->om.isIndexableDataAddrPresent()` to `TR::Compiler->om.isOffHeapAllocationEnabled()`.